src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -274,19 +276,48 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
             {
                 Requires.NotNull(operation, nameof(operation));
 
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons in the JTF enabled scenario, we want to minimize the number of Joins performed, and also
+                // keep the size of the JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a
+                // JTF.RunAsync. This requires us to not ConfigureAwait(true) on the semaphore. However, that prevents us from
+                // resuming on the correct sync context. To partially fix this, we will at least resume you on the main thread or
+                // thread pool.
+                AsyncSemaphore.Releaser releaser;
+                bool resumeOnMainThread = this.joinableTaskCollection?.Context?.IsOnMainThread == true;
```

Comment:
> == true [](start = 94, length = 8)

nit: Drop this

Reply:
I changed it up, but the type is a `bool?`, using `== true` coalesces to `bool`.

Reply:
Ah, in that case I guess what you had was fine. Sorry for the noise. You can keep it the new way too. 

---
In reply to: [201787699](https://github.com/Microsoft/vs-threading/pull/313#discussion_r201787699) [](ancestors = 201787699)